README.md
# Image Color Builder

[![pub package](https://img.shields.io/pub/v/image_color_builder.svg)](https://pub.dartlang.org/packages/image_color_builder)<a href="https://github.com/janlionly/flutter_image_color_builder"><img src="https://img.shields.io/github/stars/janlionly/flutter_image_color_builder.svg?style=flat&logo=github&colorB=deeppink&label=stars" alt="Star on Github"></a>

A Flutter package for generating the image and the image's dominant color with a given image url.

## Usage

See examples to `/example` folder.

```dart
ImageColorBuilder(
  url: 'https://picsum.photos/200',
  // url: 'assets/images/local.jpg', // support to read local image
  fit: BoxFit.cover,
  maxCachedCount: 15,
  builder: (BuildContext context, Image? image, Color? imageColor) {
    return Container(
      padding: const EdgeInsets.all(40),
      decoration: BoxDecoration(
        color: imageColor?.withOpacity(0.8) ?? Colors.red,
        borderRadius: const BorderRadius.all(Radius.circular(8)),
      ),
      child: image ?? const Text('No image?'),
    );
  },
  placeholder: (contect, url) => Image.asset(
    'assets/images/placeholder.png',
    fit: BoxFit.fill,
  ),
  errorWidget: (context, url, error) => Image.asset(
    'assets/images/error.png',
    fit: BoxFit.fill,
  ),
)
```

## Author

Visit my github: [janlionly](https://github.com/janlionly)<br>
Contact with me by email: janlionly@gmail.com

## Contribute
I would love you to contribute to **ImageColorBuilder**

## License
**ImageColorBuilder** is available under the MIT license. See the [LICENSE](https://github.com/janlionly/flutter_image_color_builder/blob/master/LICENSE) file for more info.


CHANGELOG.md
## [1.0.0] - 10 Oct 2023

* Done: Describe initial release.

## [1.0.1] - 10 Oct 2023

* Done: Support url is optional or the image get failed to Image null safety.

## [1.0.2] - 11 Oct 2023

* Feat: Support image memory caching with the key of url.

## [1.0.3] - 12 Oct 2023

* Feat: Support to read the local path image.

## [1.0.4] - 12 Oct 2023

* Feat: Support to add placeholder and error widget.



test/image_color_builder_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:image_color_builder/image_color_builder.dart';

import 'package:flutter/material.dart';

void main() {
  test('generating the image and the dominant color with a given image url', () {
    ImageColorBuilder(
      url: 'https://picsum.photos/200',
      builder: (BuildContext context, Image? image, Color? imageColor) {
        return Container(
          padding: const EdgeInsets.all(40),
          decoration: BoxDecoration(
            color: imageColor?.withOpacity(0.8) ?? Colors.red,
            borderRadius: const BorderRadius.all(Radius.circular(8)),
          ),
          child: image ?? const Text('No image?'),
        );
      },
    );
  });
}


lib/image_color_builder.dart
library image_color_builder;

import 'dart:typed_data';
import 'dart:ui';

import 'package:flutter/material.dart';
import 'package:palette_generator/palette_generator.dart';
import 'package:http/http.dart' as http;
import 'package:flutter/services.dart' show rootBundle;

// import 'dart:developer' as dev;
class _Dev {
  static void log(String msg) {
    // dev.log(msg);
  }
}

class ImageColorBuilder extends StatelessWidget {
  const ImageColorBuilder({
    required this.builder,
    this.url,
    this.fit,
    this.isCached = true, 
    this.maxCachedCount = 50, 
    this.placeholder,
    this.errorWidget,
    super.key,
  });
   /// Image URL
  final String? url;

  /// Image Box Fix
  final BoxFit? fit;

  /// Whether cached images
  final bool isCached;

  /// Cached store
  static final Map<String, List> _cachedImages = {};

  /// Max Cached Count
  final int maxCachedCount;

  /// Builder function
  final Widget Function(BuildContext context, Image? image, Color? imageColor) builder;

  /// Placeholder Widget
  final Widget Function(BuildContext contect, String? url)? placeholder;

  /// Error Widget
  final Widget Function(BuildContext context, String? url, dynamic error)? errorWidget;

  @override
  Widget build(BuildContext context) {
    if (url == null || url!.isEmpty) {
      return builder(context, null, null);
    }
    final cache = _cachedImages[url];
    if (isCached && cache != null && cache.length >= 2) {
      return builder(
        context, 
        Image.memory(cache[0], fit: fit ?? BoxFit.fill),
        cache[1],
      );
    }
    return FutureBuilder<List>(
      future: _getImageColor(),
      builder: (BuildContext context, AsyncSnapshot<List> snapshot) {
        if (!snapshot.hasData || snapshot.hasError || snapshot.data!.isEmpty || snapshot.data!.length == 1) {
          Object? error = snapshot.data?.first;

          if (error != null && errorWidget != null) {
            return errorWidget!(context, url, error);
          } else if (placeholder != null) {
            return placeholder!(context, url);
          } 
          return builder(context, null, null);
        }

        return builder(
          context, 
          (snapshot.data?.first != null ? Image.memory(snapshot.data![0], fit: fit ?? BoxFit.fill) : null), 
          (snapshot.data != null && snapshot.data!.length >= 2 ? snapshot.data![1] : null)
        );
      }
    );
  }
  
  Future<List> _getImageColor() async {
    Object? error;

    if (url != null && url!.isNotEmpty) {
      final targetUrl = url!;

      late final Codec codec;
      late final FrameInfo frameInfo;
      late final PaletteGenerator paletteGenerator;
      late Uint8List imageBytes;

      http.Response? response;

      try {
        if (targetUrl.toLowerCase().startsWith('http')) {
          response = await http.get(Uri.parse(targetUrl));
          imageBytes = response.bodyBytes;
        } else {
          imageBytes = (await rootBundle.load(targetUrl)).buffer.asUint8List();
        }
        codec = await instantiateImageCodec(imageBytes);
        frameInfo = await codec.getNextFrame();
        paletteGenerator = await PaletteGenerator.fromImage(frameInfo.image);
      } catch (e) {
        error = e;
      }

      if (error == null && (response == null || (response.statusCode == 200))) {
        if (isCached) {
          _cachedImages[targetUrl] = [
            imageBytes, 
            paletteGenerator.dominantColor?.color, 
            DateTime.now().microsecondsSinceEpoch
          ];
          _Dev.log('current images:${_cachedImageDescription()}');
          _clearHalfOfCache();
        }
        return [imageBytes, paletteGenerator.dominantColor?.color];
      }
    }

    return error != null ? [error] : [];
  }

  void _clearHalfOfCache() {
    if (_cachedImages.length >= maxCachedCount) {
      _Dev.log('origin images:${_cachedImageDescription()}');
      final sorted = _cachedImages.entries.toList()..sort((e1, e2) => e1.value[2].compareTo(e2.value[2]));
      for (var i = 0; i < maxCachedCount/2; i++) {
        _Dev.log('removed the datetime: ${sorted[i].value[2]}');
        _cachedImages.remove(sorted[i].key);
      }
      _Dev.log('updated images:${_cachedImageDescription()}');
    }
  }

  String _cachedImageDescription() {
    String description = '\n';
    for (String key in _cachedImages.keys) {
      description += '  $key: [${_cachedImages[key]![1]}, ${_cachedImages[key]![2]}],\n';
    }
    return '\n{$description}\n';
  }
}



example/lib/main.dart
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:image_color_builder/image_color_builder.dart';
import 'dart:developer' as dev;

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // TRY THIS: Try running your application with "flutter run". You'll see
        // the application has a blue toolbar. Then, without quitting the app,
        // try changing the seedColor in the colorScheme below to Colors.green
        // and then invoke "hot reload" (save your changes or press the "hot
        // reload" button in a Flutter-supported IDE, or press "r" if you used
        // the command line to start the app).
        //
        // Notice that the counter didn't reset back to zero; the application
        // state is not lost during the reload. To reset the state, use hot
        // restart instead.
        //
        // This works for code too, not just values: Most code changes can be
        // tested with just a hot reload.
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  void _incrementCounter() {
    setState(() {
      // This call to setState tells the Flutter framework that something has
      // changed in this State, which causes it to rerun the build method below
      // so that the display can reflect the updated values. If we changed
      // _counter without calling setState(), then the build method would not be
      // called again, and so nothing would appear to happen.
      // _counter++;
      _counter = (Random().nextInt(100)) % 20;
      dev.log('random int: $_counter');
    });
  }

  @override
  Widget build(BuildContext context) {
    // This method is rerun every time setState is called, for instance as done
    // by the _incrementCounter method above.
    //
    // The Flutter framework has been optimized to make rerunning build methods
    // fast, so that you can just rebuild anything that needs updating rather
    // than having to individually change instances of widgets.
    return Scaffold(
      appBar: AppBar(
        // TRY THIS: Try changing the color here to a specific color (to
        // Colors.amber, perhaps?) and trigger a hot reload to see the AppBar
        // change color while the other colors stay the same.
        backgroundColor: Theme.of(context).colorScheme.inversePrimary,
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: Text(widget.title),
      ),
      body: Center(
        child:SizedBox(
          width: 240,
          height: 240,
          child: ImageColorBuilder(
            url: 'https://picsum.photos/${_counter + 200}',
            // url: 'assets/images/local.jpg',
            fit: BoxFit.cover,
            maxCachedCount: 10,
            builder: (BuildContext context, Image? image, Color? imageColor) {
              return Container(
                padding: const EdgeInsets.all(40),
                decoration: BoxDecoration(
                  color: imageColor?.withOpacity(0.8) ?? Colors.red,
                  borderRadius: const BorderRadius.all(Radius.circular(8)),
                ),
                child: image ?? const Center(child: Text('No image?')),
              );
            },
            placeholder: (contect, url) => Image.asset(
              'assets/images/placeholder.png',
              fit: BoxFit.fill,
            ),
            errorWidget: (context, url, error) => Image.asset(
              'assets/images/error.png',
              fit: BoxFit.fill,
            ),
          ),
        )
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: const Icon(Icons.add),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}
